subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -127,12 +129,75 @@ def _get_plan_for_product_id(self, product_id: str) -> Plan:
             metadata__original_transaction_id=original_transaction_id,
         ).order_by('subscription_end').last()
 
+    def _get_active_transaction(self, transaction_id: str, original_transaction_id: str) -> SubscriptionPayment:
+        try:
+            return SubscriptionPayment.objects.get(
+                provider_codename=self.codename,
+                provider_transaction_id=transaction_id,
+                metadata__original_transaction_id=original_transaction_id,
+                status=SubscriptionPayment.Status.COMPLETED,
+            )
+        except SubscriptionPayment.MultipleObjectsReturned:
+            # This is left as a countermeasure in case the deduplication fails or the code is still "not good enough"
+            # and generates duplicates. It allows us to read a warning from sentry instead of rushing another fix.
+            logger.warning('Multiple payments found when fetching active transaction id "%s". '
+                           'Consider cleaning it up. Returning first of them.', transaction_id)
+            return SubscriptionPayment.objects.filter(
+                provider_codename=self.codename,
+                provider_transaction_id=transaction_id,
+                metadata__original_transaction_id=original_transaction_id,
+                status=SubscriptionPayment.Status.COMPLETED,
+            ).first()
+
+    def _get_or_create_payment(self,
+                               transaction_id: str,
+                               original_transaction_id: str,
+                               user: User,
+                               plan: Plan,
+                               start: datetime.datetime,
+                               end: datetime.datetime,
+                               subscription: Optional[Subscription] = None) -> SubscriptionPayment:
+        try:
+            payment, was_created = SubscriptionPayment.objects.get_or_create(
+                provider_codename=self.codename,
+                provider_transaction_id=transaction_id,
+                defaults={
+                    'status': SubscriptionPayment.Status.COMPLETED,
+                    # In-app purchase doesn't report the money.
+                    # We mark it as None to indicate we don't know how much did it cost.
+                    'amount': None,
+                    'user': user,
+                    'plan': plan,
+                    'subscription': subscription,
+                    'subscription_start': start,
+                    'subscription_end': end,
+                }
+            )
+            if was_created:
+                payment.subscription.auto_prolong = False
+                payment.subscription.save()
+                # Note: initial transaction is the one that has the same original transaction id and transaction id.
+                payment.meta = AppleInAppMetadata(original_transaction_id=original_transaction_id)
+                payment.save()
```

Comment:
Put this in defaults maybe? No need to call another `save()` then.
```python
defaults={
    ...
    metadata=AppleInAppMetadata(...).dict(),
}
```

Reply:
This relies on the knowledge about the way the `meta` property is implemented :/ But it shouldn't be changed, ever, so it should be fine. And reduction of db operations is always good.